### ISSUE TITLE
Accelerate UniformRandomWalkMetaPathWalk generator

### ISSUE THREAD

**Issue by daokunzhang** [stellargraph/stellargraph#522]
### Description

In the run() function of the UniformRandomMetaPathWalk, when selecting the next node to walk every time, we have to select the neighbouring nodes with the qualified types through the following looping:

neighbours = [
         n_node
         for n_node in neighbours
         if self.graph.node[n_node][self.node_type_attribute]
         == metapath[d]
]

which makes the generation of meta-path based random walks quite slow. Before performing meta-path based random walks, for each node we may build several neighbouring node lists and respectively store its neighbours with regards to neighbouring node types. At each time to visit next node, we just need to use the required neighbouring node list.

### User Story
As a: research  engineer,
I want: accelerate the generation of meta-path based random walks,
so that: the keras metapath2vec implementation will be speeded up.

### Done Checklist (Development)
- [ ] Assumptions of the user story met
- [ ] Produced code for required functionality
- [ ] Branch and Pull Request build on CI
- [ ] Branch and Pull Request pass unit tests on CI
- [ ] Branch and Pull Request pass integration tests on CI
- [ ] Version number reflects new status
- [ ] Peer Code Review Performed
- [ ] Code well commented
- [ ] Documentation on Google Docs
- [ ] Documentation in repo
- [ ] CHANGELOG.md updated
- [ ] Team demo
- [ ] Mini-meetup talk
- [ ] Stakeholder sign-off

### Done Checklist (Research)
- [ ] Code Review
- [ ] Documentation on Google Docs
- [ ] Documentation in repo
- [ ] Team talk
- [ ] Mini-meetup talk

### Done Checklist (Bug)
- [ ] Bug fixed
- [ ] Branch and Pull Request build on CI
- [ ] Branch and Pull Request pass unit tests on CI
- [ ] Branch and Pull Request pass integration tests on CI
- [ ] Version number reflects new status
- [ ] Peer Code Review Performed
- [ ] Code well commented
- [ ] CHANGELOG.md updated

**Comment by youph** [comment]
@daokunzhang does this issue block any other issues?

**Comment by daokunzhang** [comment]
@youph No. It might be wrapped by a new issue--metapath2vec heterogeneous network embedding.

**Comment by yvonnegh** [comment]
Kevin / Kieran working through the clashes.
May be some conflict with adjacency work / new format..
Hopefully Kieran's work will be merged by the time Kevin returns.
Need to work closely together next week.

**Comment by deepakagrawal** [comment]
Any updates on this?
